File: server/src/main/java/org/jetbrains/bsp/bazel/server/bep/BepOutput.java
package org.jetbrains.bsp.bazel.server.bep;

import com.google.common.collect.Queues;
import java.net.URI;
import java.util.Collections;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class BepOutput {
  private final Map<String, Set<String>> outputGroups;
  private final Map<String, TextProtoDepSet> textProtoFileSets;
  private final Set<String> rootTargets;

  public BepOutput(
      Map<String, Set<String>> outputGroups,
      Map<String, TextProtoDepSet> textProtoFileSets,
      Set<String> rootTargets) {
    this.outputGroups = outputGroups;
    this.textProtoFileSets = textProtoFileSets;
    this.rootTargets = rootTargets;
  }

  public Set<String> rootTargets() {
    return rootTargets;
  }

  public Set<URI> filesByOutputGroupNameTransitive(String outputGroup) {
    var rootIds = this.outputGroups.getOrDefault(outputGroup, Collections.emptySet());
    if (rootIds.isEmpty()) {
      return Collections.emptySet();
    }

    var result = new HashSet<URI>(rootIds.size());
    var toVisit = Queues.newArrayDeque(rootIds);
    var visited = new HashSet<String>(rootIds.size());

    while (!toVisit.isEmpty()) {
      var fileSetId = toVisit.remove();
      var fileSet = textProtoFileSets.get(fileSetId);

      result.addAll(fileSet.getFiles());
      visited.add(fileSetId);

      var children = fileSet.getChildren();
      children.stream().filter(child -> !visited.contains(child)).forEach(toVisit::add);
    }

    return result;
  }
}

File: server/src/main/kotlin/org/jetbrains/bsp/bazel/server/bep/BepOutput.kt
package org.jetbrains.bsp.bazel.server.bep

import org.jetbrains.bsp.bazel.server.model.Label
import com.google.common.collect.Queues
import java.nio.file.Path

class BepOutput(
    private val outputGroups: Map<String, Set<String>> = emptyMap(),
    private val textProtoFileSets: Map<String, TextProtoDepSet> = emptyMap(),
    private val rootTargets: Set<Label> = emptySet()
) {
    fun rootTargets(): Set<Label> {
        return rootTargets
    }

    fun filesByOutputGroupNameTransitive(outputGroup: String): Set<Path> {
        val rootIds = outputGroups.getOrDefault(outputGroup, emptySet())
        if (rootIds.isEmpty()) {
            return emptySet()
        }
        val result = HashSet<Path>(rootIds.size)
        val toVisit = Queues.newArrayDeque(rootIds)
        val visited = HashSet<String>(rootIds)
        while (!toVisit.isEmpty()) {
            val fileSetId = toVisit.remove()
            val fileSet = textProtoFileSets[fileSetId]
            result.addAll(fileSet!!.files)
            val children = fileSet.children
            children.asSequence()
                .filter { child: String -> !visited.contains(child) }
                .forEach { e: String -> visited.add(e); toVisit.add(e) }
        }
        return result
    }
}
